src/util.rs

```diff
@@ -8,3 +8,25 @@ pub fn leak<T>(b: Arc<T>) -> &'static T {
         &*p
     }
 }
+
+/// A dummy value that we can give to panic to indicate that we tried
+/// to acquire a poisoned lock, e.g. during flat-combining. The
+/// intention is for this to be treated specially by the panic
+/// propagation code -- in cases where the code must pick a "best"
+/// panic to propagate, it can ignore values of this type.
+pub struct PoisonPanic;
+
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct ThreadId {
+    addr: usize
+}
+
+impl ThreadId {
+    pub fn current() -> Self {
+        // The actual `ID` value is irrelevant.  We're just using its TLS
+        // address as a unique thread key, faster than a real thread-id call.
+        thread_local!{ static ID: bool = false }
+        ID.with(|id| ThreadId { addr: id as *const bool as usize })
```

Comment:
- Is this refactoring relevant to the PR?  I don't see any new callers here.
- You did not keep it `#[inline]`, which means it will probably turn into a cross-crate call from the generic bridge functions.
- The performance claim that I made about this should probably be examined, if this ever gets used in a true hotspot.  e.g. I'm not sure if Rust ever uses static TLS, in which case there's probably a call for dynamic TLS anyway.  (Your OS may vary, no warranty is expressed or implied, yada yada.)
  - But anyway, it never showed up as a hotspot in my testing, and AFAIK there's no thread id available in the standard library, so this invented id is probably still good enough.